src/components/drawing/index.js

```diff
@@ -354,6 +354,139 @@ drawing.gradient = function(sel, gd, gradientID, type, colorscale, prop) {
     fullLayout._gradientUrlQueryParts[k] = 1;
 };
 
+/**
+ * pattern: create and apply a pattern fill
+ *
+ * @param {object} sel: d3 selection to apply this pattern to
+ *     You can use `selection.call(Drawing.pattern, ...)`
+ * @param {DOM element} gd: the graph div `sel` is part of
+ * @param {string} patternID: a unique (within this plot) identifier
+ *     for this pattern, so that we don't create unnecessary definitions
+ * @param {string} bgcolor: background color for this pattern
+ * @param {string} fgcolor: foreground color for this pattern
+ * @param {number} scale: scale of this pattern
+ * @param {number} solidity: how solid lines of this pattern are
+ * @param {string} prop: the property to apply to, 'fill' or 'stroke'
+ */
+drawing.pattern = function(sel, gd, patternID, shape, bgcolor, fgcolor, scale, solidity, prop) {
+    var fullLayout = gd._fullLayout;
+    var fullID = 'p' + fullLayout._uid + '-' + patternID;
+    var baseSize = 8 * scale;
+    var width, height;
+
+    var path = '';
+    switch(shape) {
+        case '/':
+            width = baseSize * Math.sqrt(2);
+            height = baseSize * Math.sqrt(2);
+            path = 'M-1,1l2,-2' +
+                   'M0,' + height + 'L' + width + ',0' +
+                   'M' + (width - 1) + ',' + (height + 1) + 'l2,-2';
+            break;
+        case '\\':
+            width = baseSize * Math.sqrt(2);
+            height = baseSize * Math.sqrt(2);
+            path = 'M' + (width - 1) + ',-1l2,2' +
+                   'M0,0L' + width + ',' + height +
+                   'M-1,' + (height - 1) + 'l2,2';
+            break;
+        case 'x':
+            width = baseSize * Math.sqrt(2);
+            height = baseSize * Math.sqrt(2);
+            path = 'M-1,1l2,-2' +
+                   'M0,' + height + 'L' + width + ',0' +
+                   'M' + (width - 1) + ',' + (height + 1) + 'l2,-2' +
+                   'M' + (width - 1) + ',-1l2,2' +
+                   'M0,0L' + width + ',' + height +
+                   'M-1,' + (height - 1) + 'l2,2';
+            break;
+        case '|':
+            width = baseSize;
+            height = baseSize;
+            path = 'M' + (width / 2) + ',0L' + (width / 2) + ',' + height;
+            break;
+        case '-':
+            width = baseSize;
+            height = baseSize;
+            path = 'M0,' + (height / 2) + 'L' + width + ',' + (height / 2);
+            break;
+        case '+':
+            width = baseSize;
+            height = baseSize;
+            path = 'M' + (width / 2) + ',0L' + (width / 2) + ',' + height +
+                   'M0,' + (height / 2) + 'L' + width + ',' + (height / 2);
+            break;
+        case '.':
+            width = baseSize;
+            height = baseSize;
+            break;
+    }
+
+    var pattern = fullLayout._defs.select('.patterns')
+        .selectAll('#' + fullID)
+        .data([0]);
```

Comment:
In `drawing.gradient`, which it (appropriately) looks like you modeled this after, we set the selection data & element identity to a string that fully encapsulates the gradient parameters:

https://github.com/plotly/plotly.js/blob/d0ddf03850ccfc5e625da0735b6dcea8e1d61a05/src/components/drawing/index.js#L319

That way if any of these parameters change the whole thing gets dropped by `.exit()`

We should do something similar here, otherwise we'll need to be much more careful with the rest of the code below - none of the variable settings can go in `.enter().each(...)` and items that only exist conditionally, like `rects`, need logic like `.data(rectExists ? [0] : [])`

Reply:
I added as data a string that includes the parameters of the pattern.
39e35ccc02eb5d76ea7bbdc88be61a7cb1943918